arwen/contexts/metering_test.go

```diff
@@ -279,80 +282,212 @@ func TestMeteringContext_GasUsed_NoStacking(t *testing.T) {
 
-	// simulate executing another contract on top of the parent
 	childInput := &vmcommon.ContractCallInput{VMInput: vmcommon.VMInput{}}
 	childInput.GasProvided = 500
+	childInput.CallerAddr = parentInput.RecipientAddr
 
 	metering.UseGas(childInput.GasProvided)
 	parentPointsBeforeStacking := mockRuntime.GetPointsUsed()
+	require.Equal(t, 900, int(parentPointsBeforeStacking))
+	require.Equal(t, 2100, int(metering.GasLeft()))
 
-	// child execution begins
 	mockRuntime.SCAddress = []byte("child")
 	mockRuntime.SetPointsUsed(0)
 	mockRuntime.SetVMInput(&childInput.VMInput)
 	metering.PushState()
 	metering.InitStateFromContractCallInput(&childInput.VMInput)
-	require.Equal(t, uint64(500), metering.initialGasProvided)
+	require.Equal(t, 500, int(metering.initialGasProvided))
 
+	return metering, output, mockRuntime, parentInput, parentPointsBeforeStacking, childInput
+}
+
+func TestMeteringContext_GasUsed_StackOneLevel(t *testing.T) {
+	metering, output, mockRuntime, parentInput, parentPointsBeforeStacking, _ := setUpStackOneLevel(t)
+
+	// child execution begins
 	_ = metering.DeductInitialGasForExecution(make([]byte, 100))
-	require.Equal(t, uint64(399), metering.GasLeft())
+	require.Equal(t, 400, int(metering.GasLeft()))
 
 	metering.UseGas(50)
 	gasRemaining := metering.GasLeft()
-	require.Equal(t, uint64(349), gasRemaining)
+	require.Equal(t, 350, int(gasRemaining))
 
-	gasUsedByContract = metering.GasSpentByContract()
-	require.Equal(t, uint64(151), gasUsedByContract)
+	gasSpentByContract := metering.GasSpentByContract()
+	require.Equal(t, 150, int(gasSpentByContract))
+
+	metering.UpdateGasStateOnSuccess(output.outputState)
 
 	// return to the parent
+	metering.PopMergeActiveState()
 	mockRuntime.SCAddress = []byte("parent")
-	metering.PopSetActiveState()
 	mockRuntime.SetPointsUsed(parentPointsBeforeStacking)
 	mockRuntime.SetVMInput(&parentInput.VMInput)
 
 	metering.RestoreGas(gasRemaining)
 	mockRuntime.IsContractOnStack = false
-	require.Equal(t, uint64(2448), metering.GasLeft())
+	require.Equal(t, 2450, int(metering.GasLeft()))
 
-	gasUsedByContract = metering.GasSpentByContract()
-	require.Equal(t, uint64(1401), gasUsedByContract)
+	gasSpentByContract = metering.GasSpentByContract()
+	require.Equal(t, 1550, int(gasSpentByContract))
 
 	metering.UseGas(50)
-	require.Equal(t, uint64(2398), metering.GasLeft())
+	require.Equal(t, 2400, int(metering.GasLeft()))
+
+	gasSpentByContract = metering.GasSpentByContract()
+	require.Equal(t, 1600, int(gasSpentByContract))
+
+	metering.UpdateGasStateOnSuccess(output.outputState)
+
+	gasUsed := output.outputState.OutputAccounts["parent"].GasUsed
+	require.Equal(t, 1450, int(gasUsed))
+
+	gasUsed = output.outputState.OutputAccounts["child"].GasUsed
+	require.Equal(t, 150, int(gasUsed))
+
+	gasRemaining = math.SubUint64(parentInput.GasProvided, gasSpentByContract)
+	// calculate gas remaining
+
+	require.Equal(t, int(gasRemaining), int(output.GetVMOutput().GasRemaining))
```

Comment:
Do not call `output.GetVMOutput()` after calling `metering.UpdateGasStateOnSuccess()`.